File: 剑指offer/22.链表中倒数第k个节点.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def getKthFromEnd(self, head: ListNode, k: int) -> ListNode:
        p = head
        q = p
        for _ in range(k):
            q = q.next

        while q:
            q = q.next
            p = p.next

        return p

File: 剑指offer/ 22.链表中倒数第k个节点.cpp
#include <iostream>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x)
        : val(x)
        , next(NULL)
    {
    }
};
class Solution {
public:
    ListNode* getKthFromEnd(ListNode* head, int k)
    {
        ListNode* slow = head;
        ListNode* fast = head;

        for (int i = 0; i < k; i++) {
            fast = fast->next;
        }

        while (fast != NULL) {
            slow = slow->next;
            fast = fast->next;
        }

        return slow;
    }
};